platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java

```diff
@@ -142,6 +142,16 @@ final class ConcurrentArray<T> {
         tail = head;
     }
 
+    /**
+     * Create a new concurrent array with {@link #DEFAULT_ELEMENT_ARRAY_LENGTH} subarray capacity
+     * and elements from the given stream. The created array is immutable.
+     */
+    ConcurrentArray(final Stream<T> from) {
+        this();
+        from.parallel().forEach(this::add);
```

Comment:
This is very expensive: `add()` contains a `synchronized` block and involves atomic updates.
While initializing a new instance of `ConcurrentArray` we are guaranteed to have exclusive access to it so we could do without synchronization.

Reply:
I refactored this class a little bit. Current code in the synchronized section in `add()` is extracted to a new method, `addImpl()`. This new method is very straightforward and not thread safe. It can be called from the constructor above or from `add()` with proper synchronization.